File: src/main/java/me/itzisonn_/meazy/instruction/label/GotoLabelInstruction.java
package me.itzisonn_.meazy.instruction.label;

import lombok.AllArgsConstructor;
import me.itzisonn_.meazy.instruction.Instruction;
import me.itzisonn_.meazy.instruction.BytecodeBuilders;
import org.jspecify.annotations.NullMarked;

import java.lang.classfile.CodeBuilder;
import java.lang.classfile.Label;
import java.util.function.Supplier;

@NullMarked
@AllArgsConstructor
public final class GotoLabelInstruction implements Instruction {
    private final Supplier<Label> supplier;

    @Override
    public void emit(BytecodeBuilders bytecodeBuilders) {
        CodeBuilder codeBuilder = bytecodeBuilders.getCodeBuilder();
        if (codeBuilder == null) throw new RuntimeException("Code builder is null");
        codeBuilder.goto_(supplier.get());
    }
}

File: src/main/kotlin/me/itzisonn_/meazy/instruction/label/GotoLabelInstruction.kt
package me.itzisonn_.meazy.instruction.label

import me.itzisonn_.meazy.instruction.BytecodeBuilders
import me.itzisonn_.meazy.instruction.Instruction
import kotlin.uuid.Uuid

class GotoLabelInstruction(private val uuid: Uuid) : Instruction {
    override fun emit(bytecodeBuilders: BytecodeBuilders) {
        val codeBuilder = bytecodeBuilders.codeBuilder ?: throw RuntimeException("Code builder is null")
        codeBuilder.goto_(bytecodeBuilders.getLabel(uuid))
    }
}
